### PR TITLE
Wip/superm1/usb removal

### PR DESCRIPTION
**Pull request by superm1**

### CODE REVIEW

**Comment by superm1** [comment]
I think that CI failure is real even though I can't reproduce it locally yet.

File: src/fu-engine.c [hunk 1]
```diff
@@ -3087,7 +3087,28 @@ fu_engine_udev_device_remove (FuEngine *self, GUdevDevice *udev_device)
...
+
+		/* can be specified using a quirk */
+		plugin_name = fu_device_get_plugin (device);
+		if (plugin_name != NULL) {
+			g_autoptr(GError) error = NULL;
+			FuPlugin *plugin = fu_plugin_list_find_by_name (self->plugin_list,
+									plugin_name, &error);
+			if (plugin == NULL) {
+				g_warning ("failed to find specified plugin %s: %s",
+					plugin_name, error->message);
```
Comment (hughsie) [inline src/fu-engine.c]: is the indenting right here?

File: src/fu-engine.c [hunk 2]
```diff
@@ -3087,7 +3087,28 @@ fu_engine_udev_device_remove (FuEngine *self, GUdevDevice *udev_device)
...
+				return;
+			}
+			if (!fu_plugin_runner_udev_device_removed (plugin, udev_device, &error)) {
+				g_warning ("failed to remove udev device %s: %s",
+					g_udev_device_get_sysfs_path (udev_device),
```
Comment (hughsie) [inline src/fu-engine.c]: same

File: src/fu-plugin.c [hunk 3]
```diff
@@ -100,7 +100,12 @@ typedef gboolean	 (*FuPluginUsbDeviceAddedFunc)	(FuPlugin	*plugin,
 typedef gboolean	 (*FuPluginUdevDeviceAddedFunc)	(FuPlugin	*plugin,
 							 GUdevDevice	*udev_device,
 							 GError		**error);
-
+typedef gboolean	 (*FuPluginUsbDeviceRemovedFunc)(FuPlugin	*plugin,
```
Comment (hughsie) [inline src/fu-plugin.c]: Isn't this just the same as FuPluginUsbDeviceAddedFunc? Can't we just rename that to `FuPluginUdevDeviceFunc`?

File: src/fu-plugin.c [hunk 4]
```diff
@@ -510,6 +511,15 @@ fu_plugin_device_remove (FuPlugin *plugin, FuDevice *device)
 		return;
 	}
 
+	children = fu_device_get_children (device);
+	for (guint j = 0; j < children->len; j++) {
+		FuDevice *child = g_ptr_array_index (children, j);
+		g_debug ("emit removed from %s: %s",
+			fu_plugin_get_name (plugin),
+			fu_device_get_id (child));
+		g_signal_emit (plugin, signals[SIGNAL_DEVICE_REMOVED], 0, child);
```
Comment (hughsie) [inline src/fu-plugin.c]: I'm not sure about this. Maybe there are some logical problems in removing a device that the plugin still thinks is "active"? I'd feel a lot more warm and fuzzy with some self tests.

Comment (superm1) [reply src/fu-plugin.c]: Added a self test for it.

In my mind I think that's why the plugins also need to get the `usb_device_removed` or `udev_device_removed` call back.

File: src/fu-engine.c (hunk 1 above)
Comment (superm1) [reply src/fu-engine.c]: fixed

File: src/fu-plugin.c (hunk 3 above)
Comment (superm1) [reply src/fu-plugin.c]: fixed

File: src/fu-plugin.c (hunk 4 above)
Comment (hughsie) [reply src/fu-plugin.c]: The thing is, that's a signal that says "this hardware has been removed", not "the daemon has killed this object for you". I guess my concern could be mitigated by adding some comments to the fu_device_add_parent_guid function that says that when the parent is removed the child is also going to be removed. I think for most plugins that don't store a ref of the FuDevice in the plugin private data it's fine.

Comment (superm1) [reply src/fu-plugin.c]: You know I think actually the Dell plugin keeps refs and needs to be updated for this flow.  I'll add a comment to that function and fixup the Dell plugin.

**Comment by superm1** [comment]
Having tested with a WD15, those above commits do fix a segfault on device removal introduced with recentish commits in 1.1.x but there s still a problem that removal can't work since the device is registered to the UEFI plugin now through the proxy not the dell plugin.  So that might need some more thought on how to clean this up properly.

**Comment by hughsie** [comment]
Counter proposal.

1. We also emit the children when emitting the parent in `fu_device_list_emit_device_removed()` or `fu_plugin_device_remove` -- not sure which is better.
2. We run a fu_plugin_device_removed() vfunc for **all** plugins (even ones with the plugin quirk set), which would be something like `void fu_plugin_device_removed(FuPlugin, FuDevice)`

The idea would be we're telling plugins `this device something cared about has gone away` and if they care they can do something, otherwise they ignore the event. No `GError` as the plugins can't "fail" the removal, it's purely informing them.

**Comment by superm1** [comment]
So the idea would be plugins can make the decision to remove the rest of the tree if they want to?   But not all plugins necessarily would do this?

**Comment by hughsie** [comment]
No, more of a "the daemon has removed this" kind of signal.

**Comment by superm1** [comment]
> We also emit the children when emitting the parent in fu_device_list_emit_device_removed() or fu_plugin_device_remove -- not sure which is better.

I experimented with both.  `fu_device_list_emit_device_removed` is better.  Otherwise you miss out on if the reason for removal is the daemon notifying USB/UDEV removal (plugins don't call `fu_plugin_device_remove`).

I've reworked this series for your counterproposal above and made sure that the right thing is done now when a WD15 is hotplugged.  The nice thing is this means no more "recoldplug" when the dock is removed to make sure whether or not synaptics is still there since we knew we could remove it from the removal of whole parent tree.

**Comment by superm1** [comment]
Also, did test that using `fu_plugin_device_removed` from a plugin works like expected.

File: src/fu-device.c [hunk 5]
```diff
@@ -1725,6 +1725,7 @@ fu_device_incorporate (FuDevice *self, FuDevice *donor)
 	FuDevicePrivate *priv = GET_PRIVATE (self);
 	FuDevicePrivate *priv_donor = GET_PRIVATE (donor);
 	g_autoptr(GList) metadata_keys = NULL;
+	GPtrArray *parent_guids = fu_device_get_parent_guids (donor);
```
Comment (hughsie) [inline src/fu-device.c]: Normally the g_autoptr's are last in the variable list.

Comment (superm1) [reply src/fu-device.c]: Fixed

File: src/fu-plugin.c [hunk 6]
```diff
@@ -1257,6 +1257,13 @@ fu_plugin_runner_udev_device_added (FuPlugin *plugin, GUdevDevice *udev_device,
 	return TRUE;
 }
 
+void
+fu_plugin_runner_device_removed (FuPlugin *plugin, FuDevice *device)
+{
+	fu_plugin_runner_device_generic (plugin, device,
```
[the remark this answers is not in the document]
Comment (superm1) [reply src/fu-plugin.c]: I hadn't really considered that, yeah I think you're right.  I've wired up a GError that will show a warning now instead.

**Comment by superm1** [comment]
I do see a slight problem in that if a device that is removed from fu-device-list in the chain that was not USB or UDEV `fu_plugin_runner_device_removed` doesn't get called.  Not sure the right place that should be connected though.  Feel free to push/adjust this to branch if you think of the right place.

**Comment by hughsie** [comment]
> Not sure the right place that should be connected though

Agreed, it's not generic enough. We can use the :removed signal from the device-list

> Feel free to push/adjust this to branch

I've fixed up a few little things, LGTM now. If it works for you still please merge to master.

**Comment by superm1** [comment]
After testing I found that since synaptics uses a plugin cache it needs to use this signal if another plugin caused the device to go away.  Added another commit for that.

_Hypothetically_ the same thing can happen to Thunderbolt but I think it's less likely since thunderbolt will get a uevent and clear the cache at that time.

So LGTM now.

**Pull request merged**